lib/cog/command/pipeline/executor.ex

```diff
@@ -677,35 +663,160 @@ defmodule Cog.Command.Pipeline.Executor do
   ########################################################################
   # Error Handling Functions
 
-  defp send_error(error, state) when is_binary(error),
-    do: send_reply("Whoops! An error occurred. #{error}", state)
-  defp send_error(error, state) do
-    Logger.warn("The error message #{inspect error} should be in string format for displaying to the user")
-    send_reply("Whoops! An error occurred. #{inspect error}", state)
-  end
+  # Create a standard error message for pipeline-related
+  # errors. Eventually this will be templated.
+  defp user_error(error, state) do
+    id = state.id
+    started = Cog.Events.Util.ts_iso8601_utc(state.started)
+    initiator = sender_mention_name(state)
+    pipeline_text = state.request["text"]
+    error_message = format_user_error(error, state)
+    failing = case state.current_plan do
+                %Plan{}=plan ->
+                  # If it failed during execution, there will be a
+                  # plan we can point to
+                  plan
+                nil ->
+                  # If it happened during planning, there will be an
+                  # invocation to show
+                  case state.invocations do
+                    [current|_] ->
+                      current
+                    _ ->
+                      # Otherwise, it happened sometime during
+                      # parsing, and we already have the full pipeline
+                      # text to show
+                      nil
+                  end
+              end
+
+    # All error messages start with this prologue
+    base = """
+    An error has occurred.
+
+    At `#{started}`, #{initiator} initiated the following pipeline, assigned the unique ID `#{id}`:
+
+        `#{pipeline_text}`
 
-  defp send_timeout(state) do
-    command = state.current_plan.command
-    send_reply("Hmmm. The #{command} command timed out.", state)
+    """
+
+    # If the failure happened during planning or execution, we'll have
+    # a bit more information to show
+    command_text = case failing do
+                     %Ast.Invocation{} ->
+                     """
+                     The pipeline failed planning the invocation:
+
+                         `#{failing}`
+                     """
+                     %Plan{} ->
+                     """
+                     The pipeline failed executing the command:
+
+                         `#{failing.invocation_text}`
+
+                     """
+                     nil ->
+                       nil
+                   end
+
+    # Finally, the text of the specific error that occurred
+    error_text = """
+    The specific error was:
+
+        #{error_message}
+
+    """
```

Comment:
This is great. A lot better than what we were doing. Not suggesting this now, but I wonder if we could clean it up more bu leveraging FuManchu? Having templates for errors might make things more maintainable.